From b1b08a8ad359c5d1b1d292d91000b726e7cfd82f Mon Sep 17 00:00:00 2001
From: Daniel Veillard <veillard@redhat.com>
Date: Tue, 31 Jul 2012 10:53:47 +0800
Subject: [PATCH] Strengthen behaviour of the push parser in problematic
 situations
To: libvir-list@redhat.com

Implement the maximum lookahead stategy, and fix some handling
of DTD to speed up processing.

Conflicts:
	parser.c: a bit of surgery was needed to get that patch to
	apply and work in a 2.7.6 context

Signed-off-by: Daniel Veillard <veillard@redhat.com>
---
 parser.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/parser.c b/parser.c
index 16fe9de..96ca57e 100644
--- a/parser.c
+++ b/parser.c
@@ -11684,6 +11684,10 @@ not_end_of_int_subset:
 		/*
 		 * We didn't found the end of the Internal subset
 		 */
+                if (quote == 0)
+                    ctxt->checkIndex = base;
+                else
+                    ctxt->checkIndex = 0;
 #ifdef DEBUG_PUSH
 		if (next == 0)
 		    xmlGenericError(xmlGenericErrorContext,
@@ -11692,6 +11696,7 @@ not_end_of_int_subset:
 	        goto done;
 
 found_end_int_subset:
+                ctxt->checkIndex = 0;
 		xmlParseInternalSubset(ctxt);
 		ctxt->inSubset = 2;
 		if ((ctxt->sax != NULL) && (!ctxt->disableSAX) &&
@@ -11831,6 +11836,11 @@ xmlParseCheckTransition(xmlParserCtxtPtr ctxt, const char *chunk, int size) {
             return(1);
         return(0);
     }
+    if (ctxt->instate == XML_PARSER_DTD) {
+        if (memchr(chunk, ']', size) != NULL)
+            return(1);
+        return(0);
+    }
     return(1);
 }
 
@@ -11951,6 +11961,13 @@ xmldecl_done:
                                      avail - old_avail)))
             xmlParseTryOrFinish(ctxt, terminate);
     }
+    if ((ctxt->input != NULL) &&
+         (((ctxt->input->end - ctxt->input->cur) > XML_MAX_LOOKUP_LIMIT) ||
+         ((ctxt->input->cur - ctxt->input->base) > XML_MAX_LOOKUP_LIMIT)) &&
+        ((ctxt->options & XML_PARSE_HUGE) == 0)) {
+        xmlFatalErr(ctxt, XML_ERR_INTERNAL_ERROR, "Huge input lookup");
+        ctxt->instate = XML_PARSER_EOF;
+    }
     if ((ctxt->errNo != XML_ERR_OK) && (ctxt->disableSAX == 1))
         return(ctxt->errNo);
 
-- 
1.7.11.2

